### PR TITLE
Pick Next Task

### PR DESCRIPTION
**Pull request by TheCharlatan**
Based on #98 and #96 

This is an attempt of a port of the pickNextTask routine from the old edge-currency-bitcoin plugin. An attempt is made to use promises where possible and execute all networking routines through pickNextTask.

PickNextTask allows the networking and database operations especially during synching, but also during normal operation, to run completely async.

PickNextTask is triggered when the socket's pending message queue is empty enough. It may also be triggered periodically by the timeout in the main sockets queue.

To keep track of jobs to be done, a cache object is used that is passed from call to call.

The promises's catch is used to enable full recovery upon a task failure, meaning allowing the task to be run again in case of a database or network error.

A wrapper class around the promise, called Deferred, is introduced to facilitate resolve calls on the promise outside of the Promise constructor's body. Deferred is additionally templated to give some type assurances on return.

### CODE REVIEW

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 1]
```diff
@@ -54,7 +68,18 @@ export function makeUtxoEngineState(
     blockBook
   } = config
 
-  const addressesToWatch = new Set<string>()
+  // const addressesToWatch = new Set<string>()
+  const addressesToWatch: Map<string, ShortPath> = new Map<string, ShortPath>()
+  const taskCache: TaskCache = {
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:73]: Use Memlet here to get fixed memory size. Use the same top-level disklet.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 2]
```diff
@@ -66,6 +91,7 @@ export function makeUtxoEngineState(
       processor
     })
     const percent = processedCount / totalCount
+    console.log(percent)
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:94]: Remove.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 3]
```diff
@@ -166,14 +181,14 @@ export function makeUtxoEngineState(
     },
 
     async addGapLimitAddresses(addresses: string[]): Promise<void> {
-      for (const addr of addresses) {
+      for (const address of addresses) {
         await saveAddress({
           ...commonArgs,
-          scriptPubkey: walletTools.addressToScriptPubkey(addr),
+          address,
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:187]: Revert this to script pubkey.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 4]
```diff
@@ -184,44 +199,96 @@ interface CommonArgs {
...
+}
+interface TaskCache {
+  addressWatching: boolean
+  blockWatching: boolean
+  addressSubscribeCache: Map<string, ShortPath>
+  utxosCache: Map<string, AddressUtxoCacheState>
+  rawUtxosCache: Map<IAccountUTXO, RawUtxoCacheState>
+  processedUtxosCache: Map<IAddress, ProcessedUtxoCacheState>
+  transactionsCache: Map<string, AddressTransactionCacheState>
+  updateTransactionsCache: Map<string, UpdateTransactionsCacheState>
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:238]: Just use a boolean here.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 5]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
         addressIndex: getAddressCount()
       }
       const { address } = walletTools.getAddress(path)
-      const scriptPubkey = walletTools.addressToScriptPubkey(address)
       await saveAddress({
         ...args,
-        scriptPubkey,
+        address,
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:312]: Change to scriptPubkey

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 6]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+  const {
+    addressSubscribeCache,
+    utxosCache,
+    rawUtxosCache,
+    processedUtxosCache,
+    transactionsCache,
+    updateTransactionsCache
+  } = taskCache
+
+  console.log({
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:363]: Change to io

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 7]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+    if (!state.fetching) {
+      state.fetching = true
+
+      utxosCache.delete(address)
+
+      // Fetch and process address UTXOs
+      const wsTask = await processAddressUtxos({
+        ...args,
+        ...state,
+        address
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:398]: Check if this really is address.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 8]
```diff
@@ -54,7 +68,18 @@ export function makeUtxoEngineState(
     blockBook
   } = config
 
-  const addressesToWatch = new Set<string>()
+  // const addressesToWatch = new Set<string>()
+  const addressesToWatch: Map<string, ShortPath> = new Map<string, ShortPath>()
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:72]: Use Memlet here as well, review if we really need two caches.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 9]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+      .then(() => {
+        addressSubscribeCache.clear()
+      })
+      .catch(() => {
+        taskCache.addressWatching = false
+      })
+    blockBook.watchAddresses(
+      Array.from(addressesToWatch.keys()),
+      (response: INewTransactionResponse) => {
+        const path = addressesToWatch.get(response.address)
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:440]: Check if addresses have a path when adding to addressesToWatch/addressSubscribe

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 10]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+            path
+          })
+          addToTransactionCache(
+            args,
+            response.address,
+            path.format,
+            path.branch,
+            transactionsCache
+          ).catch(() => {
+            throw new Error('failed to add to transaction cache')
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:453]: Make sure you want to kill the wallet here.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 11]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+      const wsTask = await processAddressUtxos({
+        ...args,
+        ...state,
+        address
+      })
+      return wsTask
+    }
+  }
+
+  // Check if there are any addresses pending to be subscribed
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:404]: Subscribe to none-used addresses first, then fetch the transactions, then to used addresses.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 12]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+  for (const [txId, state] of updateTransactionsCache) {
+    if (!state.fetching) {
+      state.fetching = true
+      updateTransactionsCache.delete(txId)
+      return updateTransactions({ ...args, txId })
+    }
+  }
+
+  // Loop unparsed utxos, some require a network call to get the full tx data
+  for (const [utxo, state] of rawUtxosCache) {
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:472]: Move towards the top, try to get the user to be able to spend money asap, then fetch address, then tx history.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 13]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+        address: state.address,
+        utxo,
+        id: `${utxo.txid}_${utxo.vout}`
+      })
+      return wsTask
+    }
+  }
+
+  // Loop processed utxos, these are just database ops, triggers setLookAhead
+  for (const [address, state] of processedUtxosCache) {
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:489]: Move towards the top, try to get the user to be able to spend money asap, then fetch address, then tx history.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 14]
```diff
@@ -240,27 +307,256 @@ const setLookAhead = async (args: SetLookAheadArgs): Promise<void> => {
...
+    })
+  return {
+    ...transactionMessage(txId),
+    deferred
+  }
+}
+
 interface SaveAddressArgs extends CommonArgs {
-  scriptPubkey: string
+  address: string
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:551]: Check if it is address or scriptPubkey.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 15]
```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
...
-  if (firstProcess) {
-    addressesToWatch.add(address)
-    blockBook.watchAddresses(
-      Array.from(addressesToWatch),
-      async (response): Promise<void> => {
-        await setLookAhead(args)
-        await processAddress({ ...args, address: response.address })
-      }
-    )
+const processFormatAddresses = async (
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:710]: Turn this into generic function that receives args and executor, then use that every time we need to loop over all branches for a format.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 16]
```diff
@@ -184,44 +199,96 @@ interface CommonArgs {
...
   mutexor: Mutexor
 }
 
-interface OnNewBlockArgs extends CommonArgs {}
+const run = async (
+  args: CommonArgs,
+  executor: (args: FormatArgs) => unknown
+): Promise<void> => {
+  const formatsToProcess = getWalletSupportedFormats(args.walletInfo)
+  for (const format of formatsToProcess) {
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:214]: Replace with format branch from below.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 17]
```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
...
+    format,
+    changeIndex
+  })
+  for (let i = 0; i < addressCount; i++) {
+    const path: AddressPath = {
+      format,
+      changeIndex,
+      addressIndex: i
+    }
+    console.log('processing path', path)
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:745]: Remove console.log

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 18]
```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
...
+    let scriptPubkey = await processor.fetchScriptPubkeyByPath(path)
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:746]: Just pick new param.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 19]
```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
...
+    networkQueryVal,
+    processor,
+    walletTools,
+    path,
+    taskCache
+  } = args
+  const transactionsCache = taskCache.transactionsCache
 
   const scriptPubkey = walletTools.addressToScriptPubkey(address)
   const addressData = await processor.fetchAddressByScriptPubkey(scriptPubkey)
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:782]: Check if the address is null here, throw!

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 20]
```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
...
+    .then(async (value: addressResponse) => {
+      const { transactions = [], txs, unconfirmedTxs, totalPages } = value
 
-  for (const rawTx of transactions) {
-    const tx = processRawTx({ ...args, tx: rawTx })
-    await processor.saveTransaction(tx)
-  }
+      // If address is used and previously not marked as used, mark as used.
+      const used = txs > 0 || unconfirmedTxs > 0
+      if (used && !(addressData?.used ?? false) && page === 1) {
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:791]: Use the null check further above, so need to check again.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 21]
```diff
@@ -527,67 +875,77 @@ const processRawTx = (args: ProcessRawTxArgs): IProcessorTransaction => {
...
-
     const toAdd = new Set<IUTXO>()
-    const rawUtxos = await blockBook.fetchAddressUtxos(address)
-    for (const rawUtxo of rawUtxos) {
-      const id = `${rawUtxo.txid}_${rawUtxo.vout}`
-      if (currentUtxoIds.has(id)) {
-        currentUtxoIds.delete(id)
+    for (const utxo of utxos) {
+      if (currentUtxoIds.has(utxo.id)) {
+        currentUtxoIds.delete(utxo.id)
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:947]: Turn console.log into io log on line 964

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 22]
```diff
@@ -610,53 +968,129 @@ const processAddressUtxos = async (
...
     case BIP43PurposeTypeEnum.Airbitz:
     case BIP43PurposeTypeEnum.Legacy:
-      script = (await fetchTransaction({ ...args, txid: utxo.txid })).hex
       scriptType = ScriptTypeEnum.p2pkh
+
+      // Legacy UTXOs need the previous transaction hex as the script
+      // If we do not currently have it, add it to the queue to fetch it
+      tx = await processor.fetchTransaction(utxo.txid)
+      if (tx == null) {
+        const deferred = new Deferred<ITransaction>()
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:1015]: Do it everywhere: Rename deferred to transactionTask or something more descriptive.

File: src/common/utxobased/engine/makeUtxoEngineState.ts [hunk 23]
```diff
@@ -410,80 +705,133 @@ const internalGetFreshAddress = async (
...
+        // Add the address back to the cache, incrementing the page
+        transactionsCache.set(address, {
+          path,
+          networkQueryVal,
+          fetching: false,
+          page: page + 1
+        })
+      } else {
+        // Callback for when an address has been fully processed
+        args.onAddressChecked()
```
Comment (TheCharlatan) [inline src/common/utxobased/engine/makeUtxoEngineState.ts:813]: Call this when the addresses has run through all cache tasks, e.g. when it is no longer in the cache.

File: src/common/utxobased/network/BlockBook.ts [hunk 24]
```diff
@@ -198,81 +219,71 @@ export function makeBlockBook(config: BlockBookConfig): BlockBook {
...
-      },
-      opts
-    )
-
-    return await promisifyWsMessage('getAccountInfo', {
-      ...opts,
-      descriptor: address
-    })
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  ): Promise<any> {
```
Comment (TheCharlatan) [inline src/common/utxobased/network/BlockBook.ts:244]: Try typing this again once it is merged to master.

File: src/common/utxobased/network/BlockBookAPI.ts [hunk 25]
```diff
@@ -0,0 +1,87 @@
+import { EdgeTransaction } from 'edge-core-js'
+
+export interface PartialTask {
+  method: string
+  params: unknown
```
Comment (TheCharlatan) [inline src/common/utxobased/network/BlockBookAPI.ts:5]: Add cleaner and use it on return task

Comment (TheCharlatan) [reply src/common/utxobased/network/BlockBookAPI.ts:5]: Also add error message to use when cleaner failed or rejected.

File: src/common/utxobased/network/Socket.ts [hunk 26]
```diff
@@ -258,8 +271,6 @@ export function makeSocket(uri: string, config: SocketConfig): Socket {
     wakeUp()
   }
 
-  setupTimer()
```
Comment (TheCharlatan) [inline src/common/utxobased/network/Socket.ts:261]: Re-add

File: src/common/utxobased/network/Socket.ts [hunk 27]
```diff
@@ -119,24 +122,27 @@ export function makeSocket(uri: string, config: SocketConfig): Socket {
     })
 
     wakeUp()
+    setupTimer()
```
Comment (TheCharlatan) [inline src/common/utxobased/network/Socket.ts:125]: Remove and re-add in onMessage.

**Review by TheCharlatan — COMMENTED** [review]
Address comments.

**Review by thehobbit85 — CHANGES_REQUESTED** [review]
Address Comments.
First commit - resolve conflicts
Second commit - fixes
Third commit - changing names of params

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 4 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:238]: Actually this needs to be an object to preserve changing it by reference.

File: src/common/utxobased/network/BlockBook.ts [hunk 28]
```diff
@@ -198,81 +219,71 @@ export function makeBlockBook(config: BlockBookConfig): BlockBook {
...
   }
 
   async function fetchTransaction(hash: string): Promise<ITransaction> {
-    return await promisifyWsMessage('getTransaction', { txid: hash })
+    return await promisifyWsMessage(transactionMessage(hash))
   }
 
   async function broadcastTx(transaction: EdgeTransaction): Promise<void> {
-    await promisifyWsMessage('sendTransaction', { hex: transaction.signedTx })
+    await promisifyWsMessage(broadcastTxMessage(transaction))
```
[the remark this answers is not in the document]
Comment (TheCharlatan) [reply src/common/utxobased/network/BlockBook.ts:286]: Doesn't seem worth the added type complexity to me.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 8 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:72]: Defer this to a later PR.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 15 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:710]: This is changed again after rebase, so I don't think it's worth touching it again before rebase.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 16 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:214]: Same as format branch above, this is changed after rebasing on current master, so has to be fixed independently upon rebase.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 11 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:404]: This requires some change in setLookAhead by the looks of it to distinguish between used and non-used transactions. Subscribe is anyway only on an array, so not sure what benefit this brings.

File: src/common/utxobased/network/BlockBookAPI.ts (hunk 25 above)
Comment (TheCharlatan) [reply src/common/utxobased/network/BlockBookAPI.ts:5]: Can this be done in a later task? Seems to only be tangentially related to pick next task. The current code also has no cleaners.

File: src/common/utxobased/engine/makeUtxoEngineState.ts (hunk 9 above)
Comment (TheCharlatan) [reply src/common/utxobased/engine/makeUtxoEngineState.ts:440]: This seems to already be the case in both instances setting the addressSubscribeCache.  SetLookAhead uses walletTools to get the address by path, processPathAddresses, gets the addresses from getNumAddressesFromPathPartition.

**Pull request closed without merging**